### PR TITLE
[Api](ft): Make s3 routes real GET requests

### PR DESCRIPTION
**Pull request by alexandre-merle**
- Use querystring to send params for s3 get requests
- Send as the same way as IAM request (Action parameter)

To be merge with https://github.com/scality/Vault/pull/398

### CODE REVIEW

File: lib/IAMClient.js [hunk 1]
```diff
@@ -293,23 +349,23 @@ class IAMClient {
         }
 
         let ret = '';
+        if (method === 'GET') {
+            if (data['x-scal-request-uids']) {
+                options.headers = {
+                    'x-scal-request-uids': data['x-scal-request-uids'],
+                };
+                delete data['x-scal-request-uids']; // eslint-disable-line
```
Comment (jmunoznaranjo) [inline lib/IAMClient.js]: slow but necessary :(

**Comment by jmunoznaranjo** [comment]
Seems good :+1:

File: lib/IAMClient.js (hunk 1 above)
Comment (michael-zapata) [inline lib/IAMClient.js]: How is this necessary?

Comment (alexandre-merle) [inline lib/IAMClient.js]: It's not, can be remove :)

**Comment by jmunoznaranjo** [comment]
@alexandre-merle feel free to merge if I'm offline

**Comment by alexandre-merle** [comment]
Addressed :)

File: lib/IAMClient.js [hunk 2]
```diff
@@ -202,14 +201,25 @@ class IAMClient {
...
 
-        const data = { additionaldata: { stringToSign: string,
-                                         signatureFromRequest: signature,
-                                         hashAlgorithm: algo,
-                                         accessKey } };
-        if (options.reqUid !== undefined) {
-            data[requestUidKey] = options.reqUid;
-        }
-        this.request('GET', '/auth/v2', callback, data);
+        const data = {
```
Comment (LaurenSpiegel) [inline lib/IAMClient.js]: Cool.  I can use this object to send info along for the authorization (policy evaluation) request as well.

File: lib/IAMClient.js [hunk 3]
```diff
@@ -316,23 +360,22 @@ class IAMClient {
...
+        }
         const req = this.useHttps ?
             https.request(options) : http.request(options);
-
-        if (method === 'POST' && typeof data === 'object') {
+        if (method === 'POST') {
             auth.generateV4Headers(req, data,
                     this.accessKey, this.secretKeyValue, 'iam');
             req.write(queryString.stringify(data, null, null, {
                 encodeURIComponent,
```
Comment (LaurenSpiegel) [inline lib/IAMClient.js]: Why do you specify encodeURIComponent here?  I read the node docs as saying this is for changing the default.
https://nodejs.org/api/querystring.html#querystring_querystring_stringify_obj_sep_eq_options

**Comment by LaurenSpiegel** [comment]
Odd that we're sending so much data through the querystring (and it will be even more data once we are policy checking), but if needed to match what AWS CLI/SDK is doing, ok. 👍

File: lib/IAMClient.js (hunk 3 above)
Comment (alexandre-merle) [inline lib/IAMClient.js]: Default is querystring.escape() :) encodeURIComponent is the name of the parameter :)

**Comment by alexandre-merle** [comment]
@LaurenSpiegel if there is too much data send by querystring, we can change s3 request to POST without too much changes later :)

**Comment by alexandre-merle** [comment]
Addressed :)

File: lib/IAMClient.js [hunk 4]
```diff
@@ -263,11 +284,22 @@ class IAMClient {
...
+        };
+        this.request('GET', '/', (err, data, code) => {
+            if (err) {
+                return callback(err);
+            }
+            return callback(null, {
+                message: {
+                    body: data,
+                    code,
+                    message: 'attributes were retrieved',
```
Comment (rahulreddy) [inline lib/IAMClient.js]: `attributes were retrieved` => `Attributes retrieved`

File: lib/IAMClient.js [hunk 5]
```diff
@@ -283,11 +315,22 @@ class IAMClient {
...
```
Comment (rahulreddy) [inline lib/IAMClient.js]: `attributes were retrieved` => `Attributes retrieved`

File: tests/unit/getAccountPropertiesTests.js [hunk 6]
```diff
@@ -76,19 +77,24 @@ const testExpecteds = buildExpecteds();
...
     let inputArray;
 
-    if (req.method === 'GET' && req.url === '/acl/emailAddresses') {
-        inputArray = JSON.parse(req.headers.additionaldata).canonicalIds;
-    } else if (req.method === 'GET' && req.url === '/acl/canonicalIds') {
-        inputArray = JSON.parse(req.headers.additionaldata).emailAddresses;
+    if (req.method === 'GET') {
+        const index = req.url.indexOf('?');
+        const data = querystring.parse(req.url.substring(index + 1), null,
+            null, { decodeURIComponent });
```
Comment (rahulreddy) [inline tests/unit/getAccountPropertiesTests.js]: You don't have to send decodeURIComponent here

File: tests/unit/v2authTests.js [hunk 7]
```diff
@@ -74,18 +68,19 @@ function processRequest(requestObject) {
 }
 
 function handler(req, res) {
-    if (req.method === 'GET' && req.url === `/auth/v2`) {
-        const testCaseIndex = processRequest(
-            JSON.parse(req.headers.additionaldata));
+    if (req.method === 'GET') {
+        const index = req.url.indexOf('?');
+        const data = querystring.parse(req.url.substring(index + 1), null,
+            null, { decodeURIComponent });
```
Comment (rahulreddy) [inline tests/unit/v2authTests.js]: same here, you don't have to send `decodeURIComponent`

File: tests/unit/v2authTests.js [hunk 8]
```diff
@@ -114,9 +109,13 @@ describe('v2 auth tests with mockup server', () => {
                 { algo: hashAlgorithms[testIndex] },
                 (err, response) => {
                     assert.deepStrictEqual(err, expectedErrors[testIndex]);
-                    assert.deepStrictEqual(
-                        response,
-                        expectedResponseBodies[testIndex]);
+                    if (!response) {
+                        assert.deepEqual(response,
```
Comment (rahulreddy) [inline tests/unit/v2authTests.js]: I dont understand what we are testing here with the falsey value from response

**Comment by alexandre-merle** [comment]
Adressed :)

**Comment by rahulreddy** [comment]
👍

**Comment by alexandre-merle** [comment]
thx @jmunoznaranjo @LaurenSpiegel @rahulreddy 
Merging on behalf of @jmunoznaranjo

**Pull request merged**